public/app/features/query/state/runRequest.ts

```diff
@@ -159,6 +169,7 @@ export function runRequest(
       request.endTime = Date.now();
 
       state = processResponsePacket(packet, state);
+      cache.set(cacheKey, state.panelData);
```

Comment:
Cache is set after getting the result, so if there are multiple requests initiated before the cache is set they will be executed (e.g. on initial render). Another way to apporach it could be shortly debouncing runRequest per hash. Either way, as you said it may come with some perf cost so maybe it'd make sense if it was optional and enabled on case by case basis.

To cache interpolated queries maybe the cache could live closer to the [.query()](https://github.com/grafana/grafana/blob/5a8384a2455bbd3c0ba5ec67e5f5e3cc4a836904/packages/grafana-runtime/src/utils/DataSourceWithBackend.ts#L129) method. There's already queryCachingTTL for backend caching but it's only used in Enterprise.

My gut feeling is that It seems it could be a nice thing to have in general if was opt-in rather than default 🤔

Reply:
I'd advise some caution when thinking of moving it closer to the `query` method unless there is a way to opt-out of caching – some teams (including ours, the Alerting team) have already implemented RTK Query in parts of the code base but still call the `query()` function in `queryFn`. A double layer of cache would be ... well you know what they say about cache invalidation :D